### DISCUSSION TITLE
Different approach to setting temperatures

### DISCUSSION THREAD

**Discussion by dooniem** [frodeheg/no.sparegris#77] [Ideas]
**I STARTED WRITING THIS AT NIGHT IN THE EDITOR.
I SEE I CAN'T FINISH THIS RIGHT NOW, SO IN CASE I LOSE ALL TEXT SOMEHOW, I'LL PUBLISH AND FINISH THIS SUGGESTION LATER**

I own multiple Mill ovens and love their way of thinking when it comes to changing your set-points for temperature, and that you can have individual timetables on each unit, if you so please. So my inspiration comes from Mill. Let me try to explain my idea:

**<ins>Goals:</ins>**

- Update the set-point temperature in one place, not go through 8 modes/menus to update the set-point for 1 device
- Put ovens in my children's room in "night mode" earlier than the master bedroom
- Change temperatures on thermostats on a different time schedule than the rest of my house

<ins>What needs to be added in Piggy Bank to achieve this:</ins>
- Time Schedule
- Priority Schedule  (Given that people really need different priorities for different modes?)
- 3 set-point temperatures for each unit – "1, 2 and 3" or maybe "Comfort, Idle and Cold" to seperate them from exisiting modes? 

Piggy's 3+5 modes (Day, Night, Vacation + userdefined) can co-exist in this proposal. The game changer is the Time Schedule.
EDIT: After writing my thoughts out, I'm not sure it makes sense to keep the modes like they currently are...hopefully I'll edit this with a good proposal, or someone reading this will chime in?

**<ins>Time Schedule:</ins>**
On app-install, a default Time Schedule is provided => 24 hours of Comfort mode. This is assigned to all your units unless manually changed. You can create as many Time schedules as you like (or limited to X if that's better for Frode). In each Time Schedule you assign the start hour and end hour for each "Temperature state" (Comfort, Idle and Cold). You can choose to ignore Idle and/or Cold.

Here's a screenshot from the Mill app illustrating the point, just with different names for the temperature states (I thought they crashed with the existing Piggy names).
![image](https://global.discourse-cdn.com/business4/uploads/athom/optimized/3X/2/f/2f29ed7dc481c3f645582eff4cd62a385546f1ef_2_275x500.jpeg)

I don't propose you copy exactly this image in Piggy, Just the concept. To be fair I could live for years with just "Weekdays" and "Weekends".

I my case I would have the Time Schedules "Children's room",  "Thermostats", "General", "Off", "Home office".
I would then assign a Time Schedule to each unit who accepts temperature. Making the children's room warmer before the master bedroom gets warm. Adjusting thermostats on bathroom according to our daily routine. 
I have an office at home who doesn't need heating unless home office. Then I could through a favorite flow quickly change the office heater from "Off" to "Home office" before bedtime and start working in a warm home office.

**<ins>3 set-point temperatures for each unit</ins>**
I'm thinking something like this, for illustration purposes, open for better ways:
![image](https://user-images.githubusercontent.com/9369839/201786051-24c5b0ee-4fec-4f43-ad87-e28dbb75b76b.png)

This way it's much quicker to adjust your temperatures, in one place. 
Either list all temp-devices like this, or it pop's up when clicked "somewhere"?

**<ins>Price control</ins>**
My thoughts are that Comfort is the price sensitive state, where you adjust a delta according to cheap, very cheap etc.
I think "Idle" and "Cold" could be selectable for price controlling. Idle for instance would be wise to increase if you are out and it's dirt cheap, but maybe you don't want to lower the temp, unless it's crazy expensive...So no illustration yet for this :p 

**<ins>Priority Schedule</ins>**
Since the current state of Piggy allows for different prioritization for each mode, I thought my suggestion would mess this up.
So my best suggestion to keep this in line, for those who prioritize differently on each mode

**Comment by frodeheg** [comment]
The time schedule is a good idea. I would like to do this, but I also would like to make the system backward compatible so I don't break a lot of users operations, so I think I will like to do this in this way:
* Add the time schedule (single day only)
* Let you set a time schedule per mode per device
* Then your mode can now be weekend or weekday if you prefer.
* There will be no use for Comfort, Idle, and Cold as you would rather have a mode for each of those

* in order to make it backward compatible, I will do the following:
  * Automatically create empty time schedules for Each setpoint temperature that is used under all modes
  * Automatically assign every device to the new schedules that fits with the temperature set in each mode.
  * This will possibly create a lot of schedules and make a mess but it will make the app work as it did before. And users will be able to clean up the mess and create better schedules and change their mode-changing flows as they wish.

As a step two in this plan, I can also try to group all mode changes per device rather than browse through the different modes, but that will be the version after.

Priority I'll leave out of it for now as it's a completely separate feature.

How does this sound?

**Reply by dooniem** [reply]
I follow you 80%. But sounds good, easier to comment when implemented

**Comment by frodeheg** [comment]
So... I made a preview for you to play around with under the "Advanced" -> "Time Schedule" menu.

The time schedule playground will not affect anything right now, but you can try out the interface if you like and see what you think.
The missing part is that once the schedules has been created they should be assignable to the devices under mode.... and the old price control page will go away as it will be part of the schedule.

Though... right now I added one price control per schedule, do you think I should add it per timeslot instead? Or maybe a selector that lets you choose between "Price control disabled" / "Price control per schedule" / "Price control per timeslot"?

**Comment by dooniem** [comment]
Nice. I think adding all those price options would be better. I want to turn off price control at night for certain things, so a "per time slot" is deeply appreciated. Setting them all to 0 will disable, but I guess it's more user friendly to add it as an option.
Right now it's hard to set the correct hour for the time slot. I reckon that is because of all the intro text at the top. Which leads to my iOS screen scrolling instead of moving the time slot. I managed to increase by one hour at a time or 7 or so at a time. If there's room, a text box for manual input as well as the dynamic would be golden. Maybe this is different on Android?

**Comment by frodeheg** [comment]
I did some refinements to the UI + added price control per timeslot.
Not sure why your screen scrolls when you drag the timeslots (I did add `touch-action: none` in the CSS for the area covered by the timeline, so it should not). If it still does, please let me know and I'll investigate.

**Reply by dooniem** [reply]
Much better, was not able to reproduce the scroll problem. Nice touch with static temperature and no temp control. Looking good!
